### DISCUSSION TITLE
How to use the API to download CSV?

### DISCUSSION THREAD

**Discussion by rickcecil** [nocodb/nocodb#4263] [Q&A] [answered]
I am trying to automate a CSV export with NocoDB. 

For reference, I have thousands of records in a database. I am using NocoDB to identify relevant records for my research. I then want to export the IDs of these records so that I run them through another program that will generate HTML files for me. I am currently doing this manually, but it seems like I should be able to automate this using the API. I have tried a variety of things over the last few days.

In the Swagger API UI, I can run the appropriate API command and download the CSV. However, when I attempt to do this via the command line, using the command listed in the Swagger UI, it does not work. Not only is it not working, I am not getting any error messages, which makes it impossible for me to troubleshoot.

I have tried specifying an output file with -o. I have tried using " >> file.csv" at the end (not sure what this is called). And none of it is working. 

Here is the command that I am using:

```
curl -X 'GET' \
  'https://my.localdomain.com/api/v1/db/data/v1/<database>/<table>/views/<view>/export/csv' \
  -H 'accept: application/octet-stream' \
  -H 'xc-auth: <code>' \
  -H 'xc-token: <code>'
```

It is exactly the command that works in the Swagger UI. 

Any help would be appreciated -- even if it is a pointer on where to conduct further research.

Thanks!

**Comment by wingkwong** [comment]
Can you show the project info? I just tried on 0.98.3 and it's working fine.

From your command, you just need to specify either `xc-auth` or `xc-token`. You can see the difference [here](https://docs.nocodb.com/FAQs#what-is-the-difference-between-auth-token-and-api-token-).

**Comment by rickcecil** [comment]
Here's my Project Info:

Node: **v12.22.12**
Arch: **x64**
Platform: **linux**
Docker: **true**
Database: **sqlite3**
ProjectOnRootDB: **false**
RootDB: **pg**
PackageVersion: **0.98.3**

FWIW, I removed the xc-auth and am now only using xc-token (that's the API token and not the Auth Token, correct?) and it is still not working. I also generated a new API token and it still did not work.

Slight tangent, but regarding the auth vs token...So I had generated an API Token and mistakenly used that for both the auth and token in the SwaggerUI -- and it worked in both fields -- or, at least, it showed the authorization for the xc-auth as being successful. Not sure if it should have worked in the xc-auth, but it did, so I had left it in.

**Reply by wingkwong** [reply]
To your questions

1. Yes. xc-token is API token.
2. Yes. Both would work.

You may try the following

1. Try executing another GET API to see if you can see the result. (Wanna test if the project is due to the export API)
2. By running the curl command, the result could only show in the console. You have to add something like  `--output test.csv` at the end. You can double check if the file can be generated.

May I know the number of records of your table? You may ping us in Discord if you could record a video of how you reproduce the issue.

**Comment by rickcecil** [comment]
Unfortunately, the result is not displaying in the console, either. In fact, I thought that was the initial issue and spent time sorting out where exactly to put the --output. Spent some time before I realized that the reason I am getting a 0 byte csv file is because it is not pulling any data down. :(

Not sure why I didn't think to try a different GET API. Let me do that and I'll also test a different project as well. If successful, I will report back here. If not, I will jump in the Discord for additional support

There are 396,751 records in the table, but I only want to export 1,793 records (via the view).

**Comment by rickcecil** [comment, marked as answer]
Well, I am an idiot. 

I had spaces in both the names of the project and the view and did not %20 them. I saw that the spaces were not converted in the Swagger UI, but didn't think twice about them needing to be encoded. On a whim, I thought I would try that before trying a different GET call. And what do you know, it is working.

So, for anyone struggling to get the API call to work, %20 those spaces in your CLI. 

(Though, not using spaces would likely be the easier way to deal with this.)

Thanks for the assist, @wingkwong!

BTW, NocoDB is fantastic! Very much looking forward to how it evolves!
